### ISSUE TITLE
default initialization for `Conv2DCCLayer` does not make sense when dimshuffle=False

### ISSUE THREAD

**Issue by benanne** [benanne/Lasagne#49]
The default initialization for the cuda-convnet based 2D convolution (`Conv2DCCLayer`) is `init.Uniform` without any parameters, which defaults to the "Glorot"-style uniform initialization based on fan in and fan out.

If `Conv2DCCLayer` is used with `dimshuffle=False`, it uses `c01b` ordering for the input dimensions instead of the default of `bc01`. This causes the initializer to compute the fan in and fan out incorrectly: it computes the receptive field size as the product of the last two dimensions, which is correct for `bc01` but not for `c01b`.

The easiest solution is probably to define a custom `Uniform_c01b` initializer and use that as the default instead when `dimshuffle=False`. I'll try to sort that out soon, unless someone has a better idea for this.

**Comment by f0k** [comment]
> I'll try to sort that out soon, unless someone has a better idea for this.

In #87, we more or less decided to do a separate Glorot initializer. In addition to an optional `gain` argument, it could also have optional `fan_in` and `fan_out` arguments overriding the default computation of those from the shape. `Conv2DCCLayer` could compute the fan in and fan out and pass that to the `init.Glorot` then. This won't work with the current way of creating the default initializer directly in the function declaration, though... if we want to keep this, deriving a `Glorot_c01b` initializer from `Glorot` will be the easiest solution.
While we're at it, what about 1D and 3D convolution? Is everything correct there? If not, we might want to go for the `fan_in` and `fan_out` arguments after all, instead of creating a bunch of specialized `Glorot` subclasses.

**Comment by benanne** [comment]
Everything should be fine for 1D and 3D convolution for now, the fan in is just computed as the product of all trailing dimensions beyond the 1st in the bc01 case, so that works out okay.

Are you proposing to add `fan_in` and `fan_out` as optional arguments for initializers? Because for the Orthogonal initializer, I don't think this is sufficient, you actually need to pass in the set of axes that constitute the fan-in in that case.

**Comment by f0k** [comment]
> Are you proposing to add `fan_in` and `fan_out` as optional arguments for initializers? Because for the Orthogonal initializer, I don't think this is sufficient, you actually need to pass in the set of axes that constitute the fan-in in that case.

I was just thinking of Glorot-style initialization. Having `fan_in` and `fan_out` be slices or axis lists would be possible as well. If that would be needed for `Orthogonal` anyway, we could do the same for `Glorot` and not create an extra `Glorot_c01b` class. Something along the lines of:

```python
def get_fan(shape, hint):
    if isinstance(hint, int):
        # this would allow people to use Glorot with fully custom fan-in and fan-out
        # if that's not needed, we can drop get_fan and just compute fan-in and
        # fan-out directly in Glorot.sample()
        return hint
    return np.prod(np.array(shape)[hint])

class Glorot(Initializer):  # doesn't actually make sense to derive from Uniform
    def __init__(self, gain=1, fan_in=slice(0), fan_out=slice(1)):
        super(Glorot, self).__init__()
        self.gain = gain
        self.fan_in = fan_in
        self.fan_out = fan_out

    def sample(self, shape):
        fan_in = get_fan_factor(shape, self.fan_in)
        fan_out = get_fan_factor(shape, self.fan_out)
        scale = self.gain * np.sqrt(6. / (fan_in + fan_out))
        return floatX(np.random.uniform(low=-scale, high=scale, size=shape))

class Conv2DLayer(Layer):
    def __init__(self, ..., W=init.Glorot(fan_in=[1,2,3], fan_out=[0,2,3]), ...)

class Conv2DCCLayer(Layer):
    def __init__(self, ..., W=init.Glorot(fan_in=..., fan_out=...), ...)
```

Having the unparameterized `init.Glorot()` do the correct thing for standard bc01 convolution would be possible as well.

`Orthogonal` could get a similar interface, or one limited to axis lists.

**Comment by benanne** [comment]
This bug has been fixed in #180.
